### ISSUE TITLE
Remove the tf.cond and tf.control_dependencies in the codebase.

### ISSUE THREAD

**Issue by gabrieldemarmiesse** [tensorflow/addons#1794]
**Describe the feature and the current behavior/state.**

With tf.function, and more precisely autograph, tf.cond and tf.control_dependencies are not needed anymore. One can take a look at the the [RFC](https://github.com/tensorflow/community/blob/master/rfcs/20180918-functions-not-sessions-20.md) for reference. TensorFlow Addons should aim to be an example for contributors who want to know how to make custom keras objects in TF2 in a clean way. It will also help us by making the code easier to read.

**Relevant information**
- Are you willing to contribute it (yes/no): Some of it
- Are you willing to maintain it going forward? (yes/no): Some of it
- Is there a relevant academic paper? (if so, where): the [RFC](https://github.com/tensorflow/community/blob/master/rfcs/20180918-functions-not-sessions-20.md) but it's not academic :p 
- Is there already an implementation in another framework? (if so, where): no
- Was it part of tf.contrib? (if so, where): no

**Which API type would this fall under (layer, metric, optimizer, etc.)**

All

**Who will benefit with this feature?**

devs who read the code, so some users, contributors and maintainers.

**Any other info.**

I left `tf.while_loop` out of this proposal because there are some extra arguments that we might need to give to the `while_loop` function, and that's not possible with autograph.

Feel free to comment if you think it's a good/bad idea.

**Comment by bhack** [comment]
> I left tf.while_loop out of this proposal because there are some extra arguments that we might need to give to the while_loop function, and that's not possible with autograph.

Is it only in the [seq2seq decoder](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/decoder.py)?

**Comment by gabrieldemarmiesse** [comment]
Yes, it's the only place where there is a tf.cond.

**Comment by guillaumekln** [comment]
I think `tf.control_dependencies` are still required for assert ops. For example the documentation of [`tf.debugging.assert_equal`](https://www.tensorflow.org/api_docs/python/tf/debugging/assert_equal) states:

> **Returns:**
>
> Op that raises `InvalidArgumentError` if `x == y` is False. This can be used with `tf.control_dependencies` inside of `tf.functions` to block followup computation until the check has executed.

**Comment by bhack** [comment]
Are they missing this case in the [main doc reference](https://www.tensorflow.org/api_docs/python/tf/control_dependencies?version=nightly)?

> Note: In TensorFlow 2 with eager and/or Autograph, you should not require this method, as code executes in the expected order. Only use tf.control_dependencies when working with v1-style code or in a graph context such as inside Dataset.map.

**Comment by gabrieldemarmiesse** [comment]
@guillaumekln  @bhack thanks for the pointers! 


For user checks in `call` methods, I think we're going to go with this anyway: https://github.com/tensorflow/addons/issues/1417 which will allow us to do more checks and debugging at runtime. Since it's only in eager mode, we wouldn't have to use tf.control_dependencies.

**Comment by guillaumekln** [comment]
It seems `tf.control_dependencies` may not be needed for assert ops, contrary to what the documentation says. Consider the following example:

```python
import tensorflow as tf

@tf.function
def test(x):
    tf.debugging.assert_greater(x, 0)
    tf.print(x)
    return x

test(tf.constant(1))
test(tf.constant(-1))
```

It correctly raises an exception on the second call.

**Comment by bhack** [comment]
@guillaumekln I've already opened https://github.com/tensorflow/tensorflow/issues/39305 to try to make the docs homogeneous.

**Comment by gabrieldemarmiesse** [comment]
This can be closed :)
